### PR TITLE
DEPR: groupby.grouper

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Ref: https://github.com/pandas-dev/pandas/pull/55738#discussion_r1426070370

This reverts the deprecation of `grouper.reconstructed_codes` and `grouper.group_keys_seq` from #56149 in favor of deprecating all of `grouper`.

### CODE REVIEW

File: doc/source/whatsnew/v2.2.0.rst [hunk 1]
```diff
@@ -472,7 +472,7 @@ Other Deprecations
 - Deprecated strings ``H``, ``S``, ``U``, and ``N`` denoting units in :func:`to_timedelta` (:issue:`52536`)
 - Deprecated strings ``H``, ``T``, ``S``, ``L``, ``U``, and ``N`` denoting units in :class:`Timedelta` (:issue:`52536`)
 - Deprecated strings ``T``, ``S``, ``L``, ``U``, and ``N`` denoting frequencies in :class:`Minute`, :class:`Second`, :class:`Milli`, :class:`Micro`, :class:`Nano` (:issue:`52536`)
-- Deprecated the :class:`.BaseGrouper` attributes ``group_keys_seq`` and ``reconstructed_codes``; these will be removed in a future version of pandas (:issue:`56148`)
+- Deprecated the :class:`.DataFrameGroupBy.grouper` and :class:`SeriesGroupBy.grouper`; these will be removed in a future version of pandas (:issue:`56521`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.2.0.rst:475]: either `:attr:` or the word "attribute" somewhere in here?

File: pandas/tests/groupby/transform/test_transform.py [hunk 2]
```diff
@@ -1353,7 +1353,9 @@ def func(grp):
 
     # Check that the fastpath raises, see _transform_general
     obj = gb._obj_with_exclusions
-    gen = gb.grouper.get_iterator(obj, axis=gb.axis)
+    msg = "DataFrameGroupBy.grouper is deprecated"
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        gen = gb.grouper.get_iterator(obj, axis=gb.axis)
```
Comment (jbrockmendel) [inline pandas/tests/groupby/transform/test_transform.py:1358]: just change to _grouper?

File: pandas/tests/groupby/test_grouping.py [hunk 3]
```diff
@@ -814,19 +820,26 @@ def test_groupby_empty(self):
...
 
-        assert gr.grouper.group_info[2] == 0
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            assert gr.grouper.group_info[2] == 0
 
         # check name
-        assert s.groupby(s).grouper.names == ["name"]
+        msg = "SeriesGroupBy.grouper is deprecated"
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            assert s.groupby(s).grouper.names == ["name"]
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_grouping.py:842]: pretty much all of these can just change grouper->_grouper i think, as long as there are a couple that test the deprecation

File: pandas/tests/groupby/test_grouping.py [hunk 4]
```diff
@@ -814,19 +820,26 @@ def test_groupby_empty(self):
...
+            tm.assert_numpy_array_equal(
+                gr.grouper.group_info[0], np.array([], dtype=np.dtype(np.intp))
+            )
 
-        tm.assert_numpy_array_equal(
-            gr.grouper.group_info[1], np.array([], dtype=np.dtype(np.intp))
-        )
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            tm.assert_numpy_array_equal(
+                gr.grouper.group_info[1], np.array([], dtype=np.dtype(np.intp))
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_grouping.py:833]: if for some reason this doesnt change to _grouper, i have a medium-strong preference for doing as little inside the context as possible, so e.g. `gpr = gr.grouper` would go inside the context and the assert_numpy... would go in a separate line

**Comment by rhshadrach** [comment]
Thanks @jbrockmendel - changes made.

**Review by mroeschke — APPROVED** [review]
Merge when ready @jbrockmendel

**Review by jbrockmendel — APPROVED** [review]

**Comment by jbrockmendel** [comment]
thanks @rhshadrach

**Comment by jorisvandenbossche** [comment]
@rhshadrach one usage of this that I noticed in downstream (seaborn, https://github.com/mwaskom/seaborn/pull/3615, https://github.com/mwaskom/seaborn/blob/a3cb0f1b33551eb25fed1907d0abdd0237fac215/seaborn/categorical.py#L640) is to access the `result_index`. 

Do we have a good alternative for that?
One could do something like `pd.Index(grouped.indices.keys())`, but that's less robust (converts from python scalars)

I haven't followed the PR where you are refactoring `result_index`, so don't know if that will change anything (or only how it is calculated), but would we want to expose something similar on the groupby object itself as alternative?

**Comment by rhshadrach** [comment]
Using `result_index` is not currently reliable:

```
df = pd.DataFrame(
    {
        "a": pd.Categorical([1], categories=[1, 2]),
        "b": pd.Categorical([1], categories=[1, 2]),
        "c": 5,
    }
)
gb = df.groupby(["a", "b"])
print(gb.grouper.result_index)
# MultiIndex([(1, 1)], names=['a', 'b'])
print(gb.size())
# a  b
# 1  1    1
#    2    0
# 2  1    0
#    2    0
# dtype: int64
```

though #55738 will fix this in all cases I know of. An alternative is to perform a reduction and take the index from the result. I recommend `size` for this as it doesn't require as much computation as other methods.

```
size = 100_000
df = pd.DataFrame({'a': np.random.randint(0, 100, size), 'b': np.random.randint(0, 100, size), 'c': np.random.random(size)})
%timeit df.groupby(['a', 'b']).size().index
# 3.68 ms ± 6.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit df.groupby(['a', 'b'])._grouper.result_index
# 3.36 ms ± 31.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

We do not need to deprecate `grouper` for #55738, only certain grouper attributes (see #56149)

**Pull request merged**